src/Basescape/SoldierRankState.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+/*
+ * Copyright 2010-2023 OpenXcom Developers.
+ *
+ * This file is part of OpenXcom.
+ *
+ * OpenXcom is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * OpenXcom is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with OpenXcom.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "../Engine/State.h"
+#include "../SaveGame/Soldier.h"
+#include <vector>
+
+namespace OpenXcom
+{
+
+struct RankItem
+{
+	RankItem(const SoldierRank _rank, const std::string &_name, const int &_quantity, const bool _promotionAllowed)
+		: rank(_rank), name(_name), openings(_quantity), promotionAllowed(_promotionAllowed)
+	{
+	}
+	const SoldierRank rank;
+	const int openings;
+	const std::string name;
+	bool promotionAllowed;
+};
+
+class Base;
+class TextButton;
+class Window;
+class Text;
+class TextEdit;
+class TextList;
+class Armor;
+class ArrowButton;
+
+class SoldierRankState : public State
+{
+private:
+	Base *_base;
+	size_t _soldierId;
+
+	TextButton *_btnCancel;
+	Window *_window;
+	Text *_txtTitle, *_txtRank, *_txtOpening;
+	TextEdit *_btnQuickSearch;
+	TextList *_lstRank;
+	std::vector<RankItem> _ranks;
+	std::vector<int> _indices;
```

Comment:
not needed, there is no sorting or filtering

Reply:
There is filtering, but not sorting.